### PR TITLE
normalizing_eq, strict_cmp for LSRV subtags

### PR DESCRIPTION
**Pull request by robertbastian**
Continuation of #1709

### CODE REVIEW

**Review by zbraniecki — APPROVED** [review]

**Comment by robertbastian** [comment]
@zbraniecki could you please take an in-depth look at the locale canonicalizer changes? I don't want to lose any performance there.

File: components/locale_canonicalizer/src/locale_canonicalizer.rs [hunk 1]
```diff
@@ -112,32 +112,31 @@ pub struct LocaleCanonicalizer {
 }
 
 #[inline]
-fn uts35_rule_matches<I, V, L>(
+fn uts35_rule_matches<'a, I>(
```
[the remark this answers is not in the document]
Comment (robertbastian) [reply components/locale_canonicalizer/src/locale_canonicalizer.rs:115]: In line 381 this is called with the result of `str::split`, so `TinyStr::eq` is not possible.

File: components/locid/src/subtags/language.rs [hunk 2]
```diff
@@ -222,6 +222,28 @@ impl Language {
     pub fn is_empty(self) -> bool {
         self == Self::UND
     }
+
+    /// Compare this `Language` with BCP-47 bytes.
```
[the remark this answers is not in the document]
Comment (robertbastian) [reply components/locid/src/subtags/language.rs:226]: Can I not, it's a lot of work for very little gain. I should've done #1932 before this, and will definitely continue working on it now.

**Comment by robertbastian** [comment]
> My only comment is that you compare stringified bytes in uts35_matches for language subtag. It may be faster to pass Language and compare it as a newtype.

That's because of the one case that only has a raw language string which we don't want to parse

`cargo criterion --features bench`:
```
canonicalize/create     time:   [16.597 us 17.073 us 17.633 us]                                 
                        change: [-1.9008% +2.0000% +5.9820%] (p = 0.32 > 0.05)
                        No change in performance detected.
canonicalize/create+canonicalize                                                                            
                        time:   [71.854 us 72.639 us 73.404 us]
                        change: [-9.7121% -7.3968% -5.1626%] (p = 0.00 < 0.05)
                        Performance has improved.

canonicalize-noop/create                                                                             
                        time:   [2.5516 us 2.6509 us 2.7904 us]
                        change: [+1.6712% +5.7943% +10.686%] (p = 0.01 < 0.05)
                        Performance has regressed.
canonicalize-noop/create+canonicalize                                                                             
                        time:   [5.1814 us 5.2305 us 5.2817 us]
                        change: [-9.0598% -7.3219% -5.6808%] (p = 0.00 < 0.05)
                        Performance has improved.

likelysubtags/maximize  time:   [2.8717 us 2.8958 us 2.9227 us]                                    
                        change: [+0.0090% +3.3803% +7.0127%] (p = 0.05 < 0.05)
                        Change within noise threshold.

```

File: components/locale_canonicalizer/src/locale_canonicalizer.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (zbraniecki) [reply components/locale_canonicalizer/src/locale_canonicalizer.rs:115]: Line 381 is a rare case outlier when we handle variants. I would prefer to pay the price of `Language::from_str` there and have it match `Language` in all cases.

Is that ok with you?

**Comment by robertbastian** [comment]
And fall back to `UND` if parsing fails? The data is untrusted

File: components/locale_canonicalizer/src/locale_canonicalizer.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (robertbastian) [reply components/locale_canonicalizer/src/locale_canonicalizer.rs:115]: Good call

**Comment by robertbastian** [comment]
```
canonicalize/create     time:   [14.362 us 14.411 us 14.466 us]                                 
                        change: [-5.9787% -4.5198% -3.1220%] (p = 0.00 < 0.05)
                        Performance has improved.
canonicalize/create+canonicalize                                                                            
                        time:   [56.003 us 56.529 us 57.156 us]
                        change: [-27.707% -25.620% -23.660%] (p = 0.00 < 0.05)
                        Performance has improved.

canonicalize-noop/create                                                                             
                        time:   [2.4034 us 2.4183 us 2.4361 us]
                        change: [+4.2992% +5.3335% +6.5468%] (p = 0.00 < 0.05)
                        Performance has regressed.
canonicalize-noop/create+canonicalize                                                                             
                        time:   [5.0018 us 5.0774 us 5.1751 us]
                        change: [-3.6568% -1.6034% +0.4594%] (p = 0.14 > 0.05)
                        No change in performance detected.

likelysubtags/maximize  time:   [2.4489 us 2.4757 us 2.5075 us]                                    
                        change: [-3.4554% -1.1961% +0.6681%] (p = 0.29 > 0.05)
                        No change in performance detected.

```

**Review by zbraniecki — APPROVED** [review]
lgtm! thank you!

**Comment by robertbastian** [comment]
Whoopsie, looks like not continuing if the variant is not a valid language has different behaviour than assuming it's a language string! @zbraniecki which behaviour is correct?

**Comment by zbraniecki** [comment]
> Whoopsie, looks like not continuing if the variant is not a valid language has different behaviour than assuming it's a language string! @zbraniecki which behaviour is correct?

@dminor should know better, but at a glance, it seems confusing to me. Shouldn't we always skip the block - in one case because we didn't parse to Language, in another because the `str` didn't match anything? What am I missing?

**Comment by robertbastian** [comment]
Nevermind I introduced a bug in my last commit. Moved the iterator sorted check before the `.next()` call that removes the language from the front

**Review by dminor — APPROVED** [review]
lgtm!

**Review by zbraniecki — APPROVED** [review]
lgtm!

File: components/locid/src/subtags/language.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (sffc) [reply components/locid/src/subtags/language.rs:226]: These functions should be covered by tests before 1.0. If not in this PR then please do it in another PR.

**Review by sffc — COMMENTED** [review]
If it looks good to @zbraniecki and @dminor then it looks good to me

**Pull request merged**